arch/arm64/src/common/arm64_internal.h

```diff
@@ -0,0 +1,387 @@
+#  undef  USE_EARLYSERIALINIT
+#else
+#  if defined(CONFIG_LWL_CONSOLE)
+#    undef  USE_SERIALDRIVER
+#    undef  USE_EARLYSERIALINIT
+#  elif defined(CONFIG_CONSOLE_SYSLOG)
+#    undef  USE_SERIALDRIVER
+#    undef  USE_EARLYSERIALINIT
+#  else
+#    define USE_SERIALDRIVER 1
+#    define USE_EARLYSERIALINIT 1
+#  endif
+#endif
+
+/* If some other device is used as the console, then the serial driver may
+ * still be needed.  Let's assume that if the upper half serial driver is
+ * built, then the lower half will also be needed.  There is no need for
+ * the early serial initialization in this case.
+ */
+
+#if !defined(USE_SERIALDRIVER) && defined(CONFIG_STANDARD_SERIAL)
+#  define USE_SERIALDRIVER 1
+#endif
+
+/* Check if an interrupt stack size is configured */
+
+#ifndef CONFIG_ARCH_INTERRUPTSTACK
+#  define CONFIG_ARCH_INTERRUPTSTACK 0
+#endif
+
+/* If the floating point unit is present and enabled, then save the
+ * floating point registers as well as normal ARM registers.
+ */
+
+#define arm64_savestate(regs) (regs = (uint64_t *)CURRENT_REGS)
+#define arm64_restorestate(regs) (CURRENT_REGS = regs)
+
+/* This is the value used to mark the stack for subsequent stack monitoring
+ * logic.
+ */
+
+#define STACK_COLOR    0xdeaddead
+#define HEAP_COLOR     'h'
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+#ifndef __ASSEMBLY__
+typedef void (*up_vector_t)(void);
+#endif
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+#ifndef __ASSEMBLY__
+#ifdef __cplusplus
+#define EXTERN extern "C"
+extern "C"
+{
+#else
+#define EXTERN extern
+#endif
+
+/* AArch64 the stack-pointer must be 128-bit aligned */
+
+#define STACK_ALIGNMENT     16
+
+/* Stack alignment macros */
+
+#define STACK_ALIGN_MASK    (STACK_ALIGNMENT-1)
+#define STACK_ALIGN_DOWN(a) ((a) & ~STACK_ALIGN_MASK)
+#define STACK_ALIGN_UP(a)   (((a) + STACK_ALIGN_MASK) & ~STACK_ALIGN_MASK)
+
+struct __attribute__((__packed__)) thread_stack_element
+{
+  char data;
+};
+
+#define STACK_PTR_ALIGN(ptr) ROUND_DOWN((ptr), STACK_ALIGNMENT)
+
+#define THREAD_STACK_DEFINE(sym, size) \
+    struct thread_stack_element locate_data(".initstack") \
+     aligned_data(STACK_ALIGNMENT) sym[size]
+
+#define THREAD_STACK_ARRAY_DEFINE(sym, n, size) \
+    struct thread_stack_element locate_data(".initstack") \
+     aligned_data(STACK_ALIGNMENT) sym[n][size]
+
+#define THREAD_STACK_DEFINE_EXTERN(sym, size) \
+    EXTERN struct thread_stack_element locate_data(".initstack") \
+     aligned_data(STACK_ALIGNMENT) sym[size]
+
+#define THREAD_STACK_ARRAY_DEFINE_EXTERN(sym, n, size) \
+    EXTERN struct thread_stack_element locate_data(".initstack") \
+     aligned_data(STACK_ALIGNMENT) sym[n][size]
+
+#define STACK_PTR_TO_FRAME(type, ptr) \
```

Comment:
why not use container_of directly

Reply:
it's use at arm64_initialstate.c 
arm64_new_task, to allocate a strcut size from stack
only for readable purpose

Reply:
But all people understand the meaning of container_of

Reply:
Oh, I check container_of at nuttx.h, define is 
#define container_of(ptr, type, member) \
  ((type *)((uintptr_t)(ptr) - offsetof(type, member)))

but for  STACK_PTR_TO_FRAME, it's 
#define STACK_PTR_TO_FRAME(type, ptr) \
    (type *)((ptr) - **sizeof(type)**)
it's deference define